Source/WebCore/platform/graphics/avfoundation/WebAVSampleBufferListener.mm

```diff
@@ -0,0 +1,295 @@
+                ASSERT(_renderers.contains(renderer.get()));
+                if (auto client = _client.get())
+                    client->rendererDidReceiveError(renderer.get(), error.get());
+            });
+            return;
+        }
+
+        ASSERT_NOT_REACHED_WITH_MESSAGE("Unexpected kind of object while observing errorContext");
+        return;
+    }
+
+    if (context == outputObscuredDueToInsufficientExternalProtectionContext) {
+        RetainPtr layer = dynamic_objc_cast<AVSampleBufferDisplayLayer>(object, PAL::getAVSampleBufferDisplayLayerClass());
+        BOOL isObscured = [[change valueForKey:NSKeyValueChangeNewKey] boolValue];
+
+        ensureOnMainThread([self, protectedSelf = RetainPtr { self }, layer = WTFMove(layer), isObscured] {
+            ASSERT(_layers.contains(layer.get()));
+            if (auto client = _client.get())
+                client->outputObscuredDueToInsufficientExternalProtectionChanged(isObscured);
+        });
+        return;
+    }
+
+    [super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
+}
+
+- (void)layerFailedToDecode:(NSNotification *)notification
+{
+    RetainPtr layer = dynamic_objc_cast<AVSampleBufferDisplayLayer>(notification.object, PAL::getAVSampleBufferDisplayLayerClass());
+    RetainPtr error = dynamic_objc_cast<NSError>([notification.userInfo valueForKey:AVSampleBufferDisplayLayerFailedToDecodeNotificationErrorKey]);
+
+    ensureOnMainThread([self, protectedSelf = RetainPtr { self }, layer = WTFMove(layer), error = WTFMove(error)] {
+        if (!_layers.contains(layer.get()))
+            return;
+        if (auto client = _client.get())
+            client->layerDidReceiveError(layer.get(), error.get());
+    });
+}
+
+- (void)layerRequiresFlushToResumeDecodingChanged:(NSNotification *)notification
+{
+    RetainPtr layer = dynamic_objc_cast<AVSampleBufferDisplayLayer>(notification.object, PAL::getAVSampleBufferDisplayLayerClass());
+    BOOL requiresFlush = [layer requiresFlushToResumeDecoding];
+
+    ensureOnMainThread([self, protectedSelf = RetainPtr { self }, layer = WTFMove(layer), requiresFlush] {
+        if (!_layers.contains(layer.get()))
+            return;
+        if (auto client = _client.get())
+            client->layerRequiresFlushToResumeDecodingChanged(layer.get(), requiresFlush);
+    });
+}
+
+#if HAVE(AVSAMPLEBUFFERDISPLAYLAYER_READYFORDISPLAY)
+- (void)layerReadyForDisplayChanged:(NSNotification *)notification
+{
+    RetainPtr layer = dynamic_objc_cast<AVSampleBufferDisplayLayer>(notification.object, PAL::getAVSampleBufferDisplayLayerClass());
+    BOOL isReadyForDisplay = [layer isReadyForDisplay];
+
+    ensureOnMainThread([self, protectedSelf = RetainPtr { self }, layer = WTFMove(layer), isReadyForDisplay] {
+        if (!_layers.contains(layer.get()))
+            return;
+        if (auto client = _client.get())
+            client->layerReadyForDisplayChanged(layer.get(), isReadyForDisplay);
```

Comment:
I still think it would be good to protect client as nobody is ensuring client will remain valid during the whole call.
Maybe instead of protecting this inside the call, we should protect here:
`if (RefPtr client = _client.get())`.

We would have to add ref()/deref() virtual methods in the client interface.

Reply:
For now this is the same as the original code that was moved from SourceBufferPrivateAVFObjC